### PR TITLE
Add roll printing sample

### PR DESCRIPTION
**Pull request by nmuggli**
* Copied heavily from existing printing sample

* Changed functionality from existing printing sample as follows:

  * Only display printers that support roll printing
  * In table, create a row for each unique width for each printer

### CODE REVIEW

**Review by rayxu-google — APPROVED** [review]

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: api-samples/printing/roll-printing/README.md
Comment (nmuggli) [reply api-samples/printing/roll-printing/README.md:13]: I just asked in the bug what the ETA is.  I'll update this if I hear anything.

File: api-samples/printing/roll-printing/README.md
Comment (nmuggli) [reply api-samples/printing/roll-printing/README.md]: Done - thanks.

File: api-samples/printing/roll-printing/README.md
Comment (nmuggli) [reply api-samples/printing/roll-printing/README.md]: Looks like this bug was fixed, so I added a milestone target to the README.  I assume we want to wait until 120 launches to actually change the javascript?

File: api-samples/printing/roll-printing/README.md
Comment (jpmedley) [inline api-samples/printing/roll-printing/README.md:13]:
```suggestion
Before Chrome 120, `submitJob()` function throws an error when returning a promise.
```

Comment (jpmedley) [reply api-samples/printing/roll-printing/README.md:13]: If the barn door has been closed, we no longer care where the cow wandered off to.

File: api-samples/printing/README.md
Comment (jpmedley) [inline api-samples/printing/README.md:13]:
```suggestion
Before Chrome 120, `submitJob()` function throws an error when returning a promise.
```

Comment (jpmedley) [reply api-samples/printing/README.md:13]: If the barn door has been closed, we no longer care where the cow wandered off to.

File: api-samples/printing/roll-printing/README.md
Comment (jpmedley) [reply api-samples/printing/roll-printing/README.md]: I think you should change the code now. It's completely appropriate (and advisable) to change this when 120 reaches beta. That happens on November 1.

Comment (nmuggli) [reply api-samples/printing/roll-printing/README.md]: Just to clarify - are you suggesting to change the code now or wait for 120 to reach beta (Nov 1)?

Comment (jpmedley) [reply api-samples/printing/roll-printing/README.md]: I'm saying we're so close to beta, you might as well change it  now.

Comment (nmuggli) [reply api-samples/printing/roll-printing/README.md]: Gotcha - thanks!  I updated it to a promise in here and in the original printer sample.

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Thanks so much for working on this! There's a lot of code duplication here and I'm wondering if it would be better integrated in to the existing sample, maybe as a separate page or tab (both are just demonstrating different features of the printing API, after all). Any thoughts there?

**Comment by nmuggli** [comment]
> Thanks so much for working on this! There's a lot of code duplication here and I'm wondering if it would be better integrated in to the existing sample, maybe as a separate page or tab (both are just demonstrating different features of the printing API, after all). Any thoughts there?

Yeah - I'll look into that.  What do you think about getting rid of the whole roll-printing directory and then having a checkbox somewhere in the original sample to enable the roll-printing features?  Thoughts from anyone about that?

**Comment by oliverdunk** [comment]
> What do you think about getting rid of the whole roll-printing directory and then having a checkbox somewhere in the original sample to enable the roll-printing features? Thoughts from anyone about that?

That sounds great! Bonus points the more of the JS functions we can re-use with some additional parameters to control if it's roll printing or not :)

**Comment by nmuggli** [comment]
> > Thanks so much for working on this! There's a lot of code duplication here and I'm wondering if it would be better integrated in to the existing sample, maybe as a separate page or tab (both are just demonstrating different features of the printing API, after all). Any thoughts there?
> 
> This is not an application. Many samples, this one included, are written to support an article that explains a feature. Good expository writing always sticks to one concept at a time except when explaining how concepts relate to each other.
> 
> If a concept article refers to a sample that mixes uses case, it has to explain to someone with no prior knowledge of the concept how to distinguish one use case from another and hope the reader doesn't get confused. If more use cases are then added to the sample, all the places that reference it have to be evaluated to make sure the distinguishing text is correct, and, if necessary, update the text to make it correct.
> 
> In an application, deduplicating code reduces long-term maintenance. In code samples, it has exactly the opposite effect.

I see - that makes sense.  Would we want to revert back where each sample is separate?  Are you OK with that Oliver?

**Comment by oliverdunk** [comment]
> I see - that makes sense. Would we want to revert back where each sample is separate? Are you OK with that Oliver?

I haven't had a chance to fully review this, but it seems like roll printing is closer to an extra set of fields on the existing API vs. new methods or solving significantly differently use cases. With that in mind I still think this is a reasonable approach.

@jpmedley, let's sync up and make a decision together. Happy to be convinced but I want to avoid @nmuggli from needing to bounce back and forth between two things until we know what we're comfortable merging :)

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Spoke to Joe and we're aligned that keeping this as one sample makes sense - thanks for bearing with us there. Left a few comments and questions, but nothing major. Let me know once you've addressed them and I can test and merge.

File: api-samples/printing/printers.js
Comment (nmuggli) [reply api-samples/printing/printers.js:38]: It can be specified any time the printer supports that option (so, it's dependent on the printer's capabilities).  However, it only makes sense for roll printing.

File: api-samples/printing/roll-printing/printers.css
Comment (nmuggli) [reply api-samples/printing/roll-printing/printers.css]: I don't think this is relevant anymore since no new files are added.

File: api-samples/printing/printers.js
Comment (nmuggli) [reply api-samples/printing/printers.js:104]: Done

File: api-samples/printing/README.md
Comment (nmuggli) [reply api-samples/printing/README.md]: Added a link to our documentation.  Is this what you're looking for, or something else? (external reference somewhere?)

Comment (oliverdunk) [reply api-samples/printing/README.md]: If there was an external reference that did a good job of explaining what it means in practice, that might be nice. But not a blocker.

Comment (nmuggli) [reply api-samples/printing/README.md]: Most of what I found was specific to certain OEMs and their applications.  I did find [this](https://openprinting.github.io/gsoc2020/14-roll-paper-support/), which seems to talk about a proposal to add roll printing and the main challenge (unknown length of paper).  Would this be worth adding?  I don't see much else that seems applicable.

Comment (oliverdunk) [reply api-samples/printing/README.md]: Yeah, probably not. No worries, thanks for checking.

**Review by oliverdunk — APPROVED** [review]
Reviewed code and tested on ChromeOS with a fake printer that supports roll printing, and this looks good - LGTM. As a nit, given the size of the capabilities field it feels like opening a new page might be better than using the popup - but will leave that up to you.

@jpmedley, can you approve from a writing POV?

**Comment by nmuggli** [comment]
> Reviewed code and tested on ChromeOS with a fake printer that supports roll printing, and this looks good - LGTM. As a nit, given the size of the capabilities field it feels like opening a new page might be better than using the popup - but will leave that up to you.
> 
> @jpmedley, can you approve from a writing POV?

Makes sense - I switched to opening a new page vs a popup.  Thanks.

File: api-samples/printing/README.md
Comment (jpmedley) [reply api-samples/printing/README.md:11]: Is it "Roll Printers" or "Roll printers"? (Capitalization difference.) It needs to match the interface.

Comment (nmuggli) [reply api-samples/printing/README.md]: It's "Roll Printers", with capital 'P'.  (https://github.com/GoogleChrome/chrome-extensions-samples/pull/1006/files#diff-6997402aba3e3c35227f4aa921f1410ea5040a48f94e3acfd2125badfae4e58dR33)

**Pull request merged**
